edgecast/client/ec_client.go

```diff
@@ -0,0 +1,276 @@
+package client
+
+/*
+	This file contains the concrete client implementation for the EC SDK
+*/
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"strings"
+
+	"github.com/EdgeCast/ec-sdk-go/edgecast/auth"
+	"github.com/EdgeCast/ec-sdk-go/edgecast/internal/collectionhelper"
+	"github.com/EdgeCast/ec-sdk-go/edgecast/internal/jsonhelper"
+)
+
+const (
+	defaultHeaderAccept      string = "application/json"
+	defaultHeaderContentType string = "application/json"
+)
+
+// Do invokes an HTTP request with the given parameters
+func (c ECClient) SubmitRequest(params SubmitRequestParams) (*Response, error) {
+	req, err := c.reqBuilder.buildRequest(buildRequestParams{
+		method:      params.Method,
+		path:        params.Path,
+		rawBody:     params.Body,
+		queryParams: params.QueryParams,
+		pathParams:  params.PathParams,
+		userAgent:   c.config.UserAgent,
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("SubmitRequest: %v", err)
+	}
+
+	// Provides an object to be filled in when unmarshaling the API response
+	req.parsedResponse = params.ParsedResponse
+
+	c.config.Logger.Debug(
+		"[REQUEST-URI]:[%s] %s\n",
+		req.method,
+		req.url.String())
+	c.config.Logger.Debug("[REQUEST-BODY]:%v\n", req.rawBody)
+	c.config.Logger.Debug("[REQUEST-HEADERS]:%+v\n", req.headers)
+
+	resp, err := c.reqSender.sendRequest(*req)
+	if err != nil {
+		return nil, fmt.Errorf("SubmitRequest: %v", err)
+	}
+	bodyAsString, _ := jsonhelper.ConvertToJSONString(resp.Data, true)
+	c.config.Logger.Debug("[RESPONSE-BODY]:%s\n", bodyAsString)
+	return resp, nil
+}
+
+// buildRequest creates a new Request for the Edgecast API with query params,
+// adding appropriate headers
+func (eb ecRequestBuilder) buildRequest(
+	params buildRequestParams,
+) (*request, error) {
+	eb.logger.Debug("Building Request: %+v", params)
+	relativeURL, err := url.Parse(params.path)
+	if err != nil {
+		return nil,
+			fmt.Errorf("ecRequestBuilder.buildRequest: url.Parse: %v", err)
+	}
+
+	if !params.method.IsValid() {
+		return nil,
+			errors.New("ecRequestBuilder.buildRequest: invalid HTTP method")
```

Comment:
Hmm it's of limited value since HTTPMethos is an int enum type. It would print whatever arbitrary number was cast as an HTTPMethod. But it's no problem to include it.